dowhy/causal_refuters/add_unobserved_common_cause.py

```diff
@@ -197,3 +198,191 @@ def include_confounders_effect(self, new_data, kappa_t, kappa_y):
+        plt.plot( x, correlation_z_list, marker='o', markerfacecolor='red', markersize=8, color='red', linewidth=2)
+        plt.plot( x , new_metric_simulated_list, marker='o', markerfacecolor='olive', markersize=8, color='green', linewidth=2)
+        plt.savefig('newmetricplot.png')
+
+        c1_final = 0
+        c2_final = 0
+
+        ans_c1_c2 = 10
+        i = 0.05
+
+        threshold = c_star_max/0.05
+
+        while i<=threshold:
+            c2 = i
+            c1 = c_star_max/c2
+            U = []
+            d_y = list(pd.Series(standardized_residuals_y))
+            d_z = list(pd.Series(standardized_residuals_z))
+            for j in range(len(d_z)):
+                simulated_variable_mean = c1*d_y[j]+c2*d_z[j]
+                simulated_variable_stddev = 1
+                U.append(np.random.normal(simulated_variable_mean, simulated_variable_stddev, 1))
+             
+            U = np.array(U)
+            
+            #debiasing the variable U 
+            model = sm.OLS(U,X)
+            results = model.fit()
+            U = U.reshape(-1, )
+            final_U = U - results.fittedvalues.values
+
+            new_data['simulated'] = final_U
+
+            column1 = new_data['simulated']
+            column2 = given_data[self._outcome_name[0]]
+            correlation_y = column1.corr(column2)
+
+            column3 = new_data[self._treatment_name[0]]
+            correlation_z = column1.corr(column3)
+
+            new_metric_simulated = correlation_y*correlation_z
+            
+            new_metric_observed = max_correlation_with_y*max_correlation_with_z
+            if ans_c1_c2>=abs(new_metric_simulated - new_metric_observed):
+                ans_c1_c2 = abs(new_metric_simulated - new_metric_observed)
+                additional_condition = (max_correlation_with_y/max_correlation_with_z)
+                if ((c1/c2) <= (additional_condition + 0.3*additional_condition)) and ((c1/c2) >= (additional_condition - 0.3*additional_condition)): #choose minimum positive value 
+                    c1_final = c1
+                    c2_final = c2
+                
+            i = i*1.5
+
+        if c1_final!=0:
+            c1 =  c1_final 
+        else:
+            c1 = math.sqrt(c_star_max)
+        if c2_final!=0:
+            c2 =  c2_final 
+        else:
+            c2 = math.sqrt(c_star_max)
+
+        '''#closed form solution
+
+        print("c_star_max before closed form", c_star_max)
+
+        if max_correlation_with_z == -1000:
+            c2 = 0
+            c1 = c_star_max
+        else:
+            additional_condition = abs(max_correlation_with_y/max_correlation_with_z)
+            print("additional_condition", additional_condition)
+            c2 = math.sqrt(c_star_max/additional_condition)
+            c1 = c_star_max/c2'''
+
+        U = []
+        d_y = list(pd.Series(standardized_residuals_y))
```

Comment:
again this code is repeated, better to put it in a function. And even precompute the first two OLS models. You don't need to compute them more than once--they are the same.